Python/Product/VSInterpreters/Interpreter/CondaEnvironmentFactoryProvider.cs

```diff
@@ -300,36 +336,60 @@ internal class CondaInfoResult {
             public string RootPrefixFolder = null;
         }
 
+
+        internal class CondaListResult : List<CondaListItem> {
+
+        }
+
+        internal class CondaListItem {
+            [JsonProperty("version")]
+            public string Version { get; set; }
+        }
+
+        private string GetCondaExecutablePath() {
+            return _condaLocatorProvider?.FindLocator()?.CondaExecutablePath;
+        }
+
         private async Task<IReadOnlyList<PythonInterpreterInformation>> FindCondaEnvironments() {
-            var mainCondaExePath = _condaLocatorProvider?.FindLocator()?.CondaExecutablePath;
+            var mainCondaExePath = GetCondaExecutablePath();
             if (!string.IsNullOrEmpty(mainCondaExePath)) {
                 return await FindCondaEnvironments(mainCondaExePath);
             }
             return Enumerable.Empty<PythonInterpreterInformation>().ToList().AsReadOnly();
         }
 
         private async Task<IReadOnlyList<PythonInterpreterInformation>> FindCondaEnvironments(string condaPath) {
+
             // Make sure to run this on a background thread as it can take a long time.
-            var condaInfoResult = await System.Threading.Tasks.Task.Run(() => ExecuteCondaInfoAsync(condaPath));
-            if (condaInfoResult != null) {
-                // We skip the root to avoid duplicate entries, root is
+            var condaInfoResult = await Task.Run(() => ExecuteCondaInfoAsync(condaPath));
+
+            // if there is no conda info, we're done
+            if (condaInfoResult == null) {
+                return Enumerable.Empty<PythonInterpreterInformation>().ToList();
+            }
+
+            var condaEnvironments = new List<PythonInterpreterInformation>();
+            foreach (var folder in condaInfoResult.EnvironmentFolders) {
+
+                // if the folder doesn't exist, skip it
+                if (string.IsNullOrEmpty(folder) || !Directory.Exists(folder)) {
+                    continue;
+                }
+
+                // skip the root to avoid duplicate entries, root is
                 // discovered by CPythonInterpreterFactoryProvider already.
-                // Older versions of `conda info` used to not return the root.
-                return condaInfoResult.EnvironmentFolders
-                    .AsParallel()
```

Comment:
reason for not doing it in parallel? i'd double check the history for a reason why they added parallel

Reply:
Couple of reasons I did it this way.

1. The large chained linq statement is much harder to debug than vanilla code. I'm not a fan of this.
2. I'm not a fan of calling a function that does actual work (CreateEnvironmentInfo) nested in a Select statement. Again, harder to read, harder to debug.
3. The amount of performance we get back here is kind of negligible. I tested this with 10 conda environments and the results came back very quickly with no parallelism. The user is not going to have hundreds or thousands of conda environments.

Reply:
I would suggest breaking down the LINQ statement into several steps, with a named variable capturing each step - this way it's still high-level (i.e. the code focuses on *what* is being done, not *how* it's being done), and variable names serve as comments describing each step. For debugging, you can then break on each step and see what the current set of candidates is.

BTW the C# debugger in VS lets you set breakpoints on arbitrary subexpressions, although the UI is a bit fiddly. But it can be very useful when debugging call chains like these.

Reply:
That's exactly what I did, in addition to removing the parallelization because it honestly isn't needed for the number of environments there are.
I refactored the code to put the terminal cases first, inverted the conditionals to include early returns to make the code more readable. The logic is the same.

Reply:
And just because the tools let you debug subexpressions doesn't justify code like this. This is messy. The expression is too long. This is an example of using linq because you can, not because you should. It doesn't simplify anything, you've got a select doing actual work, etc...

In the refactored code, you can very easily set breakpoints and see if the environment folder doesn't exist, for example.

All my opinion, of course.

Reply:
I don't think there's any fundamental difference between the same number of expressions vs statements, aside from tooling support. Subjectively, clarity with an explicit loop is lower than LINQ sequence comprehension to me for the same reason why if/goto has less clarity than a loop - it's less structured and less abstract.

All that aside, it's kind of a moot question anyway since the code now needs to use `await`, which precludes (stock) LINQ.